### ISSUE TITLE
Streamline Existing Decorators

### ISSUE THREAD

**Issue by silkentrance** [testdeck/testdeck#314]
As of now, we support a rather ambiguous interface, e.g.

```
@suite.only('name')
class foo {

  @test.only('name')
  test() {}

  @test('name')
  @only
  test2() {}
}
```

I would like to simplify the decorators by eliminating

```
interface SuiteDecorator {
   only(...); /* removed */
   skip(...); /* removed */
   pending(...); /* removed */
}

interface TestDecorator {
   only(...); /* removed */
   skip(...); /* removed */
   pending(...); /* removed */
}
```

requiring the user to then use the decorators only, e.g.

```
@suite
@skip
class foo {
}
```

Similarly so for `timeout`, `retries`, and `slow`, e.g.

```
@test(slow(100))
test() {}
```

becomes

```
@test
@slow(100)
test() {}
```

IMHO this will make things easier to both author and read. Not to mention that we can then simplify our code.

In addition to that I would like to move the naming option to a separate decorator, e.g.

```
@suite('foo')
class Suite {

  @test('bar')
  foo() {}
}
```

becomes

```
@suite
@name('foo')
class Suite {

  @test
  @name('bar')
  foo() {}
}
```

Again, this will make things much easier to author, e.g. just outcomment the name decorator or remove it using simple editor commands.

AFAIK we also have a naming decorator for the test parameters...
So we could also combine these two, making it a single decorator.

**Comment by silkentrance** [comment]
@pana-cc Not to mention #312 where I have a hard time to make things work 😄 according to some rather simplistic standard typescript-eslint rules.

**Comment by panayot-cankov** [comment]
I’ve been thinking about cutting all non-essentials too, but my line of thought was going the other way around - instead of providing each property as a separate decorator to just stick with a single decorator and an optional options object. That would mean to delete using just ’@test’, all ’test.*’, the ‘@timeout’ and ‘@slow’ as well and support only:

’’’
function test(options: {
  name?: string,
  only?: boolean,
  skip?: boolean,
  timeout?: number,
  slow?: number,
  retries?: number
} = {}): MethodDecorator {
  //…
}
’’’

Similar for ‘@suite’.

:(

But I also really love how simple this looks:
’’’ typescript
@suite class MyClass {
  @test myTest() {
  }
}
’’’

Another thing I’ve been considering is drop the support for ‘done’ callbacks and stick to promises for async tests. It will avoid all the if-y function wrapping that is trying to create a method execution function with a single ‘done’ arg matching the method signature.

We could probably fork the project, or add a testdeck2 (given we consider the current 0.*-ish a v1) with a radically minimised API. We can also delete all the ifs checking for Symbols, symbols are everywhere.

**Comment by silkentrance** [comment]
Yes, having it this way would allow IDEs to pick up the typing and support completion as well.

I have begun prototyping the decorator only approach, but this is a much better approach.
I will look into this.

What about the params decorator, which is afaik the same as the test decorator with the additional parameters?

**Comment by silkentrance** [comment]
I do not think that we need to fork the project. Just releasing a final testdeck 1.0.0 version with the stuff we have and then make it a 2.0.0 as soon as the simplified API is available.

That way we have plenty of room for bug fixes on the 1.0.0 version, backported features even, if any.

**Comment by silkentrance** [comment]
Need to check whether `before`, `static before`, `after`, and `static after` support `slow`, `timeout`, or `retries`.

**Comment by silkentrance** [comment]
I am considering to eliminate the params decorator as well, making it an option of the test decorator.

Proposed interface

```
/* TODO new API */

export interface NameOption {
    name?: string | ((params: any) => string);
}

export interface RetryOption {
    retries?: number;
}

export interface TimeoutOption {
    timeout?: number;
}

export interface SlowOption {
    slow?: number;
}

export interface OnlyOption {
    only?: boolean;
}

export interface SkipOption {
    skip?: boolean;
}

export interface PendingOption {
    pending?: boolean;
}

export interface ParamsOption {
    params?: any;
}

export interface TestOptions extends NameOption, RetryOption, SlowOption, TimeoutOption, OnlyOption, SkipOption, PendingOption, ParamsOption {
}

export interface SuiteOptions extends NameOption, OnlyOption, SkipOption, PendingOption {
}
```

**Comment by panayot-cankov** [comment]
That is what I like for before and after `@init` and `@dispose`, very similar similar to `@test` and `@suite`.

Something like:
``` TypeScript
@suite() class WebTest {

  static server;
  
  @init() static async setupServer() {
    WebTest.server = await npm.start();
  }
  
  @test() async test() {
    await WebTest.server.navigateTo("mypage.com");
    await WebTest.server.clickALot("button");
  }
  
  @dispose({
    name: "shut down the react server",
    timeout: 2000
  })
  static async stopServer() {
    await WebTest.server?.stop()
  }
}
```

If we want it to resemble the tdd/bdd it could still be @before and @after. And it should again work on static and dynamic methods. It will also allow to have multiple before/after methods. I think most frameworks support that.

**Comment by silkentrance** [comment]
Yes, I was thinking along the same line. Having multiple set up and tear down methods sounds good to me.

Must we take into account the order of execution of these methods as well or do we execute them in the order they are being parsed/interpreted?

I must look into that, compare it to for example Junit and other such test frameworks, and see how they resolve both inheritance and ordering of these methods.

**Comment by panayot-cankov** [comment]
This is as far as I get with my typings magic.

``` TypeScript
// This requires tsconfig like:
// {
//     "compilerOptions": {
//         "experimentalDecorators": true,
//         "target": "ES2022"
//     }
// }

interface TestOptions {
    name?: string;
    timeout?: number;
    slow?: number;
}

// Remove the `| (() => Promise<P[]>)` from here...
type ParamsTestOptions<ParamsType> = { params: ParamsType[] | (() => Promise<ParamsType[]>) } & TestOptions;

function isParamsOptions<ParamsType>(options: TestOptions | ParamsTestOptions<ParamsType>): options is ParamsTestOptions<ParamsType> {
    return "params" in options;
}

// I woundn't mind deleting support for "done" callback
type DoneCallback = (err?) => void;

type SyncTest = (() => void)
type CallbackAsyncTest = (done: DoneCallback) => void;
type PromiseAsyncTest = (() => Promise<void>);
type TestMethodSignature = SyncTest | CallbackAsyncTest | PromiseAsyncTest;

type ParamsSyncTest<ParamsType> = ((args: ParamsType) => void);
type ParamsCallbackAsyncTest<ParamsType> = ((args: ParamsType, done: DoneCallback) => void);
type ParamsPromiseAsyncTest<ParamsType> = ((args: ParamsType) => Promise<void>);
type ParamsTestMethodSignature<ParamsType> = ParamsSyncTest<ParamsType> | ParamsCallbackAsyncTest<ParamsType> | ParamsPromiseAsyncTest<ParamsType>;

export const testsSymbol: unique symbol = Symbol("tests");
interface Prototype extends Object {
    [testsSymbol]?: {
        [key: string | symbol]: TestOptions
    }
}

// Probably add beforeMethodsSymbol and afterMethodsSymbol

type TypedMethodDecorator<T> = (target: Object, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<T>) => void;
function test(options?: TestOptions): TypedMethodDecorator<TestMethodSignature>
function test<ParamsType>(options: ParamsTestOptions<ParamsType>): TypedMethodDecorator<ParamsTestMethodSignature<ParamsType>> 
function test<ParamsType>(options?: TestOptions | ParamsTestOptions<ParamsType>): TypedMethodDecorator<TestMethodSignature> | TypedMethodDecorator<ParamsTestMethodSignature<ParamsType>> {
    return (target: Prototype, propertyKey: string | symbol) => {
        // Here you should pretty much attach the options (and params) metadata on the class somehow, because the "suite" decorator
        // executes after all test decorators

        // We used to do something like: target[propertyKey][testSymbol] = { options, params }
        // but that was attaching the metadata on the method function and wouldnt allow polymorphism (using a @test on base class to run tests in each derived class marked with @suite)
        // and also if other mutating decorators were used they would swap the method function and we couldn't get our metadata when executing @suite

        // Something like this, but if we support inheritance it may need to be a kind of Object.hasOwnProperty(target, testSymbol)... because the base class may already contain a test method with the same name... in @suite we can inspect the inheritance chain and allow options in derived classes to override options in base classes etc.
        if (!target[testsSymbol]) target[testsSymbol] = {};
        const classTests = target[testsSymbol];
        classTests[propertyKey] = options;

        if (isParamsOptions(options)) {
            // This will declare multiple tests when the @suite() decorator executes
        } else {
            // This will declare a single test
        }
    };
}

declare function describe(name: string, s: () => void);
declare function it(name: string, t: () => void);

// and then in @suite something along
function suite(): ClassDecorator {
    return <TFunction extends Function>(target: TFunction) => {

        // Here use the abstracted testdeck API instead of describe/it directly?
        // Or if the code is cure enough, we can just support multiple implementations in parallel,
        // without reusable code between the different test runners...

        describe("...", () => {
            const classTests = (target.prototype as Prototype)[testsSymbol];
            for(const propertyKey in classTests) {
                const options = classTests[propertyKey];
                // test may be a symbol here... maybe use ?.toString()
                it(options.name ?? propertyKey, () => {
                });
            }
        });
    }
}

// Attmpt 2
function test2(options?: TestOptions): TypedMethodDecorator<TestMethodSignature>
function test2<ParamsType>(options: TestOptions, ... params: ParamsType[]): TypedMethodDecorator<ParamsTestMethodSignature<ParamsType>>
function test2<ParamsType>(options?: TestOptions, ... params: ParamsType[]): TypedMethodDecorator<TestMethodSignature> | TypedMethodDecorator<ParamsTestMethodSignature<ParamsType>> {
    return (target: Object, propertyKey: string | symbol) => {
        // Here you should pretty much attach the options (and params) metadata on the class somehow, because the "suite" decorator
        // executes after all test decorators
        
        if (params && params.length > 0) {
            // This will declare multiple tests when the @suite() decorator executes
        } else {
            // This will declare a single test
        }
    };
}

interface Point {
    x: number;
    y: number;
}

class X {
    @test({
        name: "Hello World!"
    })
    myTest() {
    }

    @test()
    myCallbackAsyncTest(done: DoneCallback) {
        done(new Error("Fail fast!"));
    }

    @test({
        name: "Fancy name"
    })
    myFancyNameTest() {
    }

    @test({
        params: [
            { x: 10, y: 20 },
            { x: 22, y: 33 },
        ]
    })
    myParamsTest1({ x, y }: { x: number, y: number }) {
    }

    @test({
        params: [
            { x: 10, /* If you hit ctrl + space here, you get no suggestions :( */y: 30 },
            { x: 10, /* If you hit ctrl + space here, you get y? suggested because of the previous array item has property Y */ }
            // You get no errors for missing "y" property
        ]
    })
    myParamsTest2({ x, y }: { x: number, y: number }, done: DoneCallback) {
    }

    // This will generate a nice error! Somewhere along the lines containing: Type '{ x: number; y: string; }' is not assignable to type '{ x: number; y: number; }'.
    @test({
        params: [
            { x: 10, y: "myproperty" },
        ]
    })
    myParamsTest3({ x, y }: { x: number, y: number }) {
    }

    // This will generate an error, somewhere along the lines containing: Type '{ x: number; y: string; }' is not assignable to type '{ x: number; y: number; }'.
    @test({
        params: [
            { x: 10, y: "myproperty" }
        ]
    })
    myParamsTest4({ x, y }) {
        // x and y here are any :( need to provide a type in the signature
    }

    @test({
        // Not sure if this is at all possible with the test frameworks... probably not, but I wish it was... It could asynchronously load CSV files and declare tests if 
        // the "suite" function could be async... Like in a `describe("suite", async () => { for await(const line of loadCSVAsync()) { it("test line" + line, () => { test(line) })} })`
        async params() {
            return [
                { x: 10, y: 30}
            ]
        }
    })
    async myParamsTest5({ x, y }) {
        // x and y here are any :( need to provide a type in the signature
    }

    @test<Point>({
        params: [
            { x: 10, y: 20},
            // You again get zero help here...
        ]
    })
    async myPramsTest6({x, y}: Point) {
    }

    @test2({
        name: "MyFancyName"
    })
    async myTest2() {
    }

    @test2<Point>({},
        { x: 10, y: 20 },
        { x: 10, }, // <- You get a nice "Property y is missing... error"
        { x: 20, y: "asd" })  // <- You get a nice "string is not a number" here when you fix the above missing y
    async myTest3({x, y}: Point) {
    }
}
```

**Comment by panayot-cankov** [comment]
On a second thought we may put the test metadata in an array instead of map, so tests could preserve their state:
```
interface Prototype extends Object {
    [testsSymbol]?: {
        [key: string | symbol]: TestOptions
    }
}
```
to become something like:
```
interface Prototype extends Object {
    [testsSymbol]?: {
      propertyKey: string | symbol,
      options: TestOptions,
    }[];
}
```

**Comment by silkentrance** [comment]
@pana-cc TLDR;-) This sounds very complicated to me. I was going for making the overall API, both internal and external, to be both more simple to use and understand.

However, along with that I would like to make the `@params` decorator something that one can use multiple times, along with specific test naming, e.g.

```
@suite
class foo {

  @test
  @params({name: (params: any) => `params[0] ...`}, [params...])
  @params({pending: true, name: (params: any) => `params[0] ...`}, [params...])
  test(params: any) {}
}
```

where one can have multiple `@params` decorators on a test method, each with their own naming scheme,
alongside optional pending/only/skip etc.

similar to `@before/@after` on the class level.

**Comment by silkentrance** [comment]
@pana-cc I would further like to change the overall behaviour so that the test instance is created once only, keeping its state across all test cases unless beforeEach/afterEach modifies that state.

The current approach, where the test instance is recreated before each test and then torn down after each test makes the beforeEach and afterEach rather useless as no state can be kept between tests.

**Comment by panayot-cankov** [comment]
My perspective is test specific state is stored in instance variables and methods, and state shared between tests is stored in static variables and methods.

[xUnit.net](https://xunit.net/docs/shared-context) I think works that way.

[MSTest](https://learn.microsoft.com/en-gb/archive/blogs/nnaderi/that-pesky-mstest-execution-ordering) too has static class initialiser and cleanup, but creates a class instance per test.

What I would love to do is have static test methods added as feature. Within static methods in TypeScript you can use ‘this’ to access static properties. So it will be a user choice whether to use instance methods with clean test context, or static test methods with shared test context.

Another thing we could do - add an option in the suite whether to construct a clean new instance for each test or to share a single class instance between each test. I think the code won’t be that iff-y. The instance will either be created and destroyed beforeEach/afterEach or beforeAll/afterAll.

**Comment by panayot-cankov** [comment]
Right, I gave up trying to strongly type the @test with params , we can keep the @test relatively clean and do the @params in a separate decorator.

**Comment by panayot-cankov** [comment]
BTW, the properties on @test do not overload too much the API. We could use ‘@test’ for vanilla tests, and something like ‘@test.params(options, paramsArr)’ that way the params naming function can’t be defined along a name in test.

**Comment by silkentrance** [comment]
@pana-cc I am currently looking into suite/test execution modifiers, i.e. `only`, `skip`, and `pending`.

Having these as options in the suite options or test options will require the user to

```
@suite({only: true|condition})
or
@test({only: true|condition})
```

whereas

```
@suite({name:...})
@only

or

@test({name:...})
@params(...)
@skip(condition)
```

seems to be more intuitive and much easier to use.

**Comment by silkentrance** [comment]
Closing in favor of #315
